lib/gcov/ukgcov.c

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: MIT */
+/* Copyright © 2005-2020 Rich Felker, et al.
+ * Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the MIT License (the "License", see COPYING.md).
+ * You may not use this file except in compliance with the License.
+ */
+/*
+ * The code in this file was derived from gcc-13-onlinedocs:
+ * Source: https://gcc.gnu.org/onlinedocs/gcc/Freestanding-Environments.html
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include "ukgcov.h"
+
+#ifdef GCOV_OPT_OUTPUT_BINARY_MEMORY
+static unsigned char __attribute__ ((used)) *gcov_output_buffer;
+static __attribute__ ((used)) size_t gcov_output_buffer_pos;
+static size_t gcov_output_buffer_size = INITIAL_BUFFER_SIZE;
+#endif
+
+/* This function shall produce a reliable in order byte stream to transfer the
+   gcov information from the target to the host system.  */
+static void dump(const void *d, unsigned n, void *arg)
+{
+#ifdef GCOV_OPT_OUTPUT_BINARY_MEMORY
+    (void)arg;
+    const unsigned char *c = d;
+    if (gcov_output_buffer_pos + n >= gcov_output_buffer_size) {
+        unsigned char *temp_buffer = (unsigned char *) realloc(gcov_output_buffer, gcov_output_buffer_size <<= 1);
+
+        if (temp_buffer == NULL) {
+            free(gcov_output_buffer);
+            return;
+        }
+        gcov_output_buffer = temp_buffer;
+    }
+    memcpy(gcov_output_buffer + gcov_output_buffer_pos, c, n);
+    gcov_output_buffer_pos += n;
+#else /* GCOV_OPT_OUTPUT_SERIAL_HEXDUMP + GCOV_OPT_OUTPUT_BINARY_FILE*/
+	FILE* file = arg;
+	const unsigned char *c = d;
+	fwrite(c, n, 1, file);
+
+#endif /* GCOV_OPT_OUTPUT_BINARY_MEMORY */
+}
+
+/* The filename is serialized to a gcfn data stream by the
+   __gcov_filename_to_gcfn() function.  The gcfn data is used by the
+   "merge-stream" subcommand of the "gcov-tool" to figure out the filename
+   associated with the gcov information. */
+
+static void filename(const char *f, void *arg)
+{
+	__gcov_filename_to_gcfn(f, dump, arg);
+}
+
+/* The __gcov_info_to_gcda() function may have to allocate memory under
+   certain conditions.  */
+static void *allocate(unsigned length, void *arg)
+{
+	(void)arg;
+	return malloc(length);
+}
+
+int dump_gcov_info (void)
+{
+    const struct gcov_info *const *info = __gcov_info_start;
+	const struct gcov_info *const *end = __gcov_info_end;
+
+	/* Obfuscate variable to prevent compiler optimizations.  */
```

Comment:
Does GCC optimize this out? If so, a more elegant way is to add `__used` in the declarations above. See `include/uk/essentials.h`